Vendoring libnetwork 83743db8ceb2bdbfa0960d9a54ed2f98df4ea846

mavenugo opened the pull request:
This PR brings in 
- Native multi-host Networking using overlay tunneling (vxlan to begin with)
- Daemon flag --kv-store to configure libkv provider required for multi-host distributed states.
- Adjusted ReleaseNetwork to the new CNM API that provides LeaveAll API.

Native multihost networking can be properly exercised only under Experimental channel as all the
relevant UI `docker network`, `docker service` and `--publish-service` UI are all under experimental tag.

mavenugo committed:
Removed the default "bridge" in builder

Now that the default network mode is "default" and this mode is chosen
even if the mode is empty string, it is not correct to have builder
still pointing to "bridge" as default (though this is daemon default).

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
Adding container to secondary network to support port mapping

With publish-service and default-network support, a container could be
connected to a user-defined network that is backed by any driver/plugin.
But if the user uses port mapping or expose commands, the expectation
for that container is to behave like existing bridge network.
Thanks to the Libnetwork's CNM model, containers can be connected
to the bridge network as a secondary network in addition to the
user-specified network.

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
Vendoring in libnetwork for native multihost networking

- brings in vxlan based native multihost networking
- added a daemon flag required by libkv for dist kv operations
- moved the daemon flags to experimental

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
Reworked ReleaseNetwork to make use of libnetwork's new LeaveAll API

Signed-off-by: Madhu Venugopal <madhu@docker.com>

estesp commented:
FYI @tiborvass @jfrazelle @jhowardmsft @vbatts and others who've been wondering/looking..this fixes the issue I found last night that was causing all the CI failures in docker-py `TestRemoveLink`.  Specifically, there was a `defer` ordering problem in libnetwork's `endpoint.go` which meant it was many times dereferencing a nil pointer and crashing the daemon, causing the docker-py tests to timeout and fail.

vbatts commented:
Phil, I owe a big ole hug
On Jun 21, 2015 07:47, "Phil Estes" <notifications@github.com> wrote:

> FYI @tiborvass <https://github.com/tiborvass> @jfrazelle
> <https://github.com/jfrazelle> @jhowardmsft
> <https://github.com/jhowardmsft> @vbatts <https://github.com/vbatts> and
> others who've been wondering/looking..this fixes the issue I found last
> night that was causing all the CI failures in docker-py TestRemoveLink.
> Specifically, there was a defer ordering problem in libnetwork's
> endpoint.go which meant it was many times dereferencing a nil pointer and
> crashing the daemon, causing the docker-py tests to timeout and fail.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/14071#issuecomment-113910809>.
>

icecrime commented:
@mavenugo The PR has no docs, can you please tell us more about `--kv-store` (or add the docs if possible)? Also I'm surprised it's not experimental: does it mean that multi-host networking is ready for mainstream, but not the `docker network`/`docker service` UX just yet?

mavenugo commented:
@icecrime good point on experimental for --kv-store. I will fix that.
Since we are controlling the UX under experimental, the multi-host networking cannot be exercised with the existing non-experimental version.

runcom commented on daemon/container_linux.go:
```diff
@@ -793,12 +807,29 @@ func (container *Container) AllocateNetwork() error {
+		return err
+	}
+
+	return nil
```
nit, this could be `return container.WriteHostConfig()` w/o the previous `if`

mavenugo commented:
@icecrime @runcom addressed your comments. Moved the flags `--default-network` and `--kv-store` under experimental flag. and wrote appropriate docs.
Also added tests for --default-network. But tests for `--kv-store` requires a richer integration framework for multi-host networking and it will be taken up soon in subsequent PRs.

moxiegirl commented:
LGTM

icecrime commented:
LGTM

calavera commented:
LGTM
